plugins/synaptics-cape/fu-synaptics-cape-device.c

```diff
@@ -0,0 +1,648 @@
+#include "fu-synaptics-cape-firmware.h"
+
+#define FU_SYNAPTICS_CAPE_DEVICE_USB_CMD_WRITE_TIMEOUT	20000 /* us */
+#define FU_SYNAPTICS_CAPE_DEVICE_USB_CMD_READ_TIMEOUT	30000 /* us */
+#define FU_SYNAPTICS_CAPE_DEVICE_USB_CMD_RETRY_INTERVAL 10    /* ms */
+#define FU_SYNAPTICS_CAPE_DEVICE_USB_CMD_RETRY_TIMEOUT	300   /* ms */
+#define FU_SYNAPTICS_CAPE_DEVICE_USB_RESET_DELAY_MS	300   /* ms */
+
+#define FU_SYNAPTICS_CAPE_DEVICE_GOLEM_REPORT_ID 1
+
+#define FU_SYNAPTICS_CAPE_CMD_MAX_DATA_LEN    13 /* number of guint32*/
+#define FU_SYNAPTICS_CAPE_CMD_WRITE_DATAL_LEN 8	 /* number of guint32*/
+#define FU_SYNAPTICS_CAPE_WORD_IN_BYTES	      4	 /*bytes*/
+
+#define FU_SYNAPTICS_CAPE_CMD_APP_ID(a, b, c, d)                                                   \
+	((((a)-0x20) << 8) | (((b)-0x20) << 14) | (((c)-0x20) << 20) | (((d)-0x20) << 26))
+
+/*return codes*/
+
+#define FU_SYNAPTICS_CAPE_MODULE_RC_GENERIC_FAILURE	(-1025)
+#define FU_SYNAPTICS_CAPE_MODULE_RC_ALREADY_EXISTS	(-1026)
+#define FU_SYNAPTICS_CAPE_MODULE_RC_NULL_APP_POINTER	(-1027)
+#define FU_SYNAPTICS_CAPE_MODULE_RC_NULL_MODULE_POINTER (-1028)
+#define FU_SYNAPTICS_CAPE_MODULE_RC_NULL_STREAM_POINTER (-1029)
+#define FU_SYNAPTICS_CAPE_MODULE_RC_NULL_POINTER	(-1030)
+
+#define FU_SYNAPTICS_CAPE_MODULE_RC_BAD_APP_ID		   (-1031)
+#define FU_SYNAPTICS_CAPE_MODULE_RC_MODULE_TYPE_HAS_NO_API (-1034)
+#define FU_SYNAPTICS_CAPE_MODULE_RC_BAD_MAGIC_NUMBER	   (-1052)
+#define FU_SYNAPTICS_CAPE_MODULE_RC_CMD_MODE_UNSUPPORTED   (-1056)
+
+/* CAPE HID message structure. Little endian.*/
+typedef struct __attribute__((packed)) {
+	gint16 data_len : 16; /*number of dwords*/
+	guint16 cmd_id : 15;  /*command id*/
+	guint16 reply : 1;    /*Host want a reply from device, 1 = true*/
+	guint32 module_id;    /*module id*/
+	guint32 data[FU_SYNAPTICS_CAPE_CMD_MAX_DATA_LEN]; /*Command data*/
+} FuCapCmd;
+
+/* CAPE HID report structure.*/
+typedef struct __attribute__((packed)) {
+	guint16 report_id; /*two bytes of report id, this should be 1*/
+	FuCapCmd cmd;
+} FuCapCmdHidReport;
+
+/* CAPE Commands */
+typedef enum {
+	FU_SYNAPTICS_CMD_FW_UPDATE_START = 0xC8,	  /* notifies firmware update started*/
+	FU_SYNAPTICS_CMD_FW_UPDATE_WRITE = 0xC9,	  /* updates firmware data */
+	FU_SYNAPTICS_CMD_FW_UPDATE_END = 0xCA,		  /* notifies firmware update finished*/
+	FU_SYNAPTICS_CMD_MCU_SOFT_RESET = 0xAF,		  /* reset device*/
+	FU_SYNAPTICS_CMD_FW_GET_ACTIVE_PARTITION = 0x1CF, /* gets cur active partition number*/
+	FU_SYNAPTICS_CMD_GET_VERSION = 0x103,		  /* gets cur firmware version*/
+} FuCommand;
+
+/*GET flag*/
+#define FU_SYNAPTICS_CMD_GET_FLAG 0x100
+
+/* CAPE Fuupd device structure*/
+struct _FuSynapticsCapeDevice {
+	FuHidDevice parent_instance;
+	guint32 active_partition; /*active partition, 0 or 1*/
+	gchar *version;
+};
+
+G_DEFINE_TYPE(FuSynapticsCapeDevice, fu_synaptics_cape_device, FU_TYPE_HID_DEVICE)
+
+static gboolean
+fu_synaptics_cape_device_sendcmd_ex(FuSynapticsCapeDevice *self,
```

Comment:
since this is "new" code, why not just declare this function first rather than the directive?

Reply:
Sorry, I can't understand what you mean. what 'directive' you are referring to?

Reply:
sorry, I can't understand what you mean. Are you referring to ```fu_synaptics_cape_device_sendcmd_ex``` Could you explain more?

Reply:
No need to have a prototype if you just put the function before it's first use.